### PR TITLE
Remove memory allocation for ELF sections

### PR DESCRIPTION
**Pull request by LambertWSJ**
Since the section headers of ELF are continuous, all headers can be accessed by getting the starting address, so there is no need to allocate memory for this.

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
Eliminate the warnings reported by Codacy/Cppcheck.

**Comment by jserv** [comment]
> Eliminate the warnings reported by Codacy/Cppcheck.

Reported by Cppcheck:
```
tools/rv_histogram.c:192:31: style: Parameter 'ir' can be declared as pointer to const [constParameterPointer]
void reg_hist_incr(rv_insn_t *ir)
                              ^
tools/rv_histogram.c:217:32: style: Parameter 'ir' can be declared as pointer to const. However it seems that 'insn_hist_incr' is a callback function, if 'ir' is declared with const you might also need to cast function pointer(s). [constParameterCallback]
void insn_hist_incr(rv_insn_t *ir)
                               ^
tools/rv_histogram.c:235:46: note: You might need to cast the function pointer here
    hist_record = show_reg ? reg_hist_incr : insn_hist_incr;
                                             ^
tools/rv_histogram.c:217:32: note: Parameter 'ir' can be declared as pointer to const
void insn_hist_incr(rv_insn_t *ir)
                               ^
tools/rv_histogram.c:253:28: style: Variable 'shdr' can be declared as pointer to const [constVariablePointer]
        struct Elf32_Shdr *shdr = (struct Elf32_Shdr *) shdrs + i;
                           ^
tools/rv_histogram.c:260:20: style: Variable 'exec_end_addr' can be declared as pointer to const [constVariablePointer]
        uintptr_t *exec_end_addr =
                   ^
```

@LambertWSJ, you may submit pull requests to eliminate the warnings reported by Cppcheck later.

File: tools/rv_histogram.c [hunk 1]
```diff
@@ -189,39 +194,38 @@ void print_hist_stats(const rv_hist_t *stats, size_t stats_size)
     }
 }
 
-void reg_hist_incr(rv_insn_t *ir)
+static void reg_hist_incr(const rv_insn_t *ir)
 {
-    if (!ir)
-        return;
+    if (ir != NULL) {
```
Comment (jserv) [inline tools/rv_histogram.c]: Don't change the function body.

File: tools/rv_histogram.c [hunk 2]
```diff
@@ -241,39 +245,35 @@
...
             if ((*((uint32_t *) ptr) & FC_OPCODE) != 0x3) {
                 insn = *((uint16_t *) ptr);
-                ptr = (uintptr_t *) ((uint8_t *) ptr + 2);
+                ptr = ptr + 2;
                 goto decode;
             }
 #endif
             insn = *((uint32_t *) ptr);
-            ptr = (uintptr_t *) ((uint8_t *) ptr + 4);
+            ptr = ptr + 4;
```
[the remark this answers is not in the document]
Comment (jserv) [reply tools/rv_histogram.c]: Rewrite as `ptr += 4`.

File: tools/rv_histogram.c (hunk 1 above)
Comment (jserv) [reply tools/rv_histogram.c]: Check https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md carefully: Avoid unnecessary nesting levels.

File: tools/rv_histogram.c [hunk 3]
```diff
@@ -179,8 +183,9 @@ void print_hist_stats(const rv_hist_t *stats, size_t stats_size)
         freq = stats[i].freq;
 
         percent = ((float) freq / total_freq) * 100;
-        if (percent < 1.00)
+        if (percent < 1.00) {
```
Comment (jserv) [inline tools/rv_histogram.c]: Don't change this.

File: tools/rv_histogram.c (hunk 1 above)
Comment (LambertWSJ) [reply tools/rv_histogram.c]: Modifications similar to the single-line if statement have been restored.

File: tools/rv_histogram.c [hunk 4]
```diff
@@ -241,39 +239,37 @@
...
     bool res;
 
     for (int i = 0; i < hdr->e_shnum; i++) {
-        struct Elf32_Shdr *shdr = (struct Elf32_Shdr *) shdrs + i;
+        const struct Elf32_Shdr *shdr = (const struct Elf32_Shdr *) &shdrs[i];
+        bool is_prg = shdr->sh_type & SHT_PROGBITS;
+        bool has_insn = shdr->sh_flags & SHF_EXECINSTR;
 
-        if (!(shdr->sh_type & SHT_PROGBITS && shdr->sh_flags & SHF_EXECINSTR))
+        if (!(is_prg && has_insn))
```
[the remark this answers is not in the document]
Comment (LambertWSJ) [reply tools/rv_histogram.c:261]: @jserv 
For modifying a single line if statement, should I follow MISRA rules or [CONTRIBUTING.md](https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md)?

File: tools/rv_histogram.c [hunk 5]
```diff
@@ -145,34 +141,36 @@
             filename);
 }
 
-static bool parse_args(int argc, char **args)
+static bool parse_args(int argc, char *args[])
```
[the remark this answers is not in the document]
Comment (LambertWSJ) [reply tools/rv_histogram.c]: @jserv 
The compiler will issue a warning if `args` is declared as a const array, Should I also change the `args` of the main function to `const char *args[]` in order to pass misra's rules?

File: tools/rv_histogram.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jserv) [reply tools/rv_histogram.c:261]: Follow [CONTRIBUTING.md](https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md).

File: tools/rv_histogram.c [hunk 6]
```diff
@@ -241,39 +239,37 @@ int main(int argc, char *args[])
...
             if ((*((uint32_t *) ptr) & FC_OPCODE) != 0x3) {
                 insn = *((uint16_t *) ptr);
-                ptr = (uintptr_t *) ((uint8_t *) ptr + 2);
+                ptr += 2;
                 goto decode;
             }
 #endif
             insn = *((uint32_t *) ptr);
-            ptr = (uintptr_t *) ((uint8_t *) ptr + 4);
+            ptr += 4;
```
Comment (jserv) [inline tools/rv_histogram.c:278]: Can you confirm if `ptr += 4` is equivalent to the original form?

File: tools/rv_histogram.c [hunk 7]
```diff
@@ -241,39 +239,37 @@ int main(int argc, char *args[])
...
+        const uint8_t *exec_end_addr = &exec_start_addr[shdr->sh_size];
+        uint8_t *ptr = exec_start_addr;
         uint32_t insn;
 
         while (ptr < exec_end_addr) {
 #if RV32_HAS(EXT_C)
             if ((*((uint32_t *) ptr) & FC_OPCODE) != 0x3) {
                 insn = *((uint16_t *) ptr);
-                ptr = (uintptr_t *) ((uint8_t *) ptr + 2);
+                ptr += 2;
```
Comment (jserv) [inline tools/rv_histogram.c:273]: Can you confirm if `ptr += 2` is equivalent to the original form?

File: tools/rv_histogram.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jserv) [reply tools/rv_histogram.c]: > The compiler will issue a warning if `args` is declared as a const array, Should I also change the `args` of the main function to `const char *args[]` in order to pass MISRA's rules?

I think so.

**Review by henrybear327 — APPROVED** [review]

File: tools/rv_histogram.c (hunk 6 above)
Comment (LambertWSJ) [reply tools/rv_histogram.c:278]: Theoretically, the results of the two are the same. By checking the number of loop executions and the histogram of the build/*.elf files, it is confirmed that they are the same as the original code.

BTW, `cc.elf` execution failed (core dump).

File: tools/rv_histogram.c (hunk 7 above)
Comment (LambertWSJ) [reply tools/rv_histogram.c:273]: ditto

File: tools/rv_histogram.c (hunk 6 above)
Comment (jserv) [reply tools/rv_histogram.c:278]: > BTW, `cc.elf` execution failed (core dump).

Create a new issue on GitHub for tracking.

**Pull request merged**